Imagej prep

jburel opened the pull request:
In this PR,
* Refactor import code to handle imagePlus object.
* Add 2 entries to the OMERO plugin menu.

To Test:
 * make sure you can still import images using the standard build.

ImageJ Test
 * Add the omero_imageJ to the plugin. 
 * Add the BF-tools jar. If you have it installed, you will need to update it.

test 1:
* Open an image in ImageJ
* Under the plugin menu. Select OMERO>Upload and Save To OMERO.
* The data manager and the importer dialog should pop up.
* Add the file to the queue and import

test 2:
* Open an image in ImageJ. Crop it for example
* Under the plugin menu. Select OMERO>Upload and Save To OMERO.
* The data manager and the importer dialog should pop up.
* Add the file to the queue and import. An ome-tiff will be generated and imported.


ROIs are not yet supported for test 1

jburel commented:
This is still work in progress. It will allow @pwalczysko @gusferguson @bramalingam  to check status.

pwalczysko commented:
test 1:

  - ``the data manager and importer should pop up`` -> only the importer popped up, there was a short blink just like if something "wanted" to pop up prior to this, but it did not...

pwalczysko commented:
Ah, the data manager did pop up, after a long time (in the meantime, I already used the importer and clicked ``Add to the queue`` button (which worked fine)

pwalczysko commented:
I can add the image to the queue many times over. The importer will import all of them, but will report the import of only one (the new "import1" tab will have just one item in the list). When checking in the Data Manager, all the images in queue were imported though.

pwalczysko commented:
BUG: The import using the normal importer (not opened via ImageJ, but straight from  Insight) fails with ``test_images_good/bd-pathway`` plate. 

```
java.lang.RuntimeException: Failure response on import!
Category: ::omero::grid::ImportRequest
Name: error-on-init
Parameters: {message=For input string: "-merge-deploy/ManagedRepository/user-4_5/2015-01/12/11-36-01", stacktrace=java.lang.NumberFormatException: For input string: "-merge-deploy/ManagedRepository/user-4_5/2015-01/12/11-36-01"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.lang.Integer.parseInt(Integer.java:492)
	at java.lang.Integer.parseInt(Integer.java:527)
	at loci.formats.in.BDReader.initFile(BDReader.java:422)
	at loci.formats.FormatReader.setId(FormatReader.java:1404)
	at loci.formats.ImageReader.setId(ImageReader.java:835)
	at loci.formats.ReaderWrapper.setId(ReaderWrapper.java:651)
	at loci.formats.ChannelFiller.setId(ChannelFiller.java:223)
	at loci.formats.ReaderWrapper.setId(ReaderWrapper.java:651)
	at loci.formats.ChannelSeparator.setId(ChannelSeparator.java:289)
	at loci.formats.ReaderWrapper.setId(ReaderWrapper.java:651)
	at loci.formats.Memoizer.setId(Memoizer.java:627)
	at loci.formats.ReaderWrapper.setId(ReaderWrapper.java:651)
	at ome.services.blitz.repo.CheckedPath.bfSetId(CheckedPath.java:403)
	at ome.services.blitz.repo.ManagedImportRequestI.open(ManagedImportRequestI.java:620)
	at ome.services.blitz.repo.ManagedImportRequestI.init(ManagedImportRequestI.java:225)
	at omero.cmd.HandleI.steps(HandleI.java:427)
	at omero.cmd.HandleI$1.doWork(HandleI.java:365)
	at omero.cmd.HandleI$1.doWork(HandleI.java:361)
	at sun.reflect.GeneratedMethodAccessor293.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at ome.services.util.Executor$Impl$Interceptor.invoke(Executor.java:576)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.security.basic.EventHandler.invoke(EventHandler.java:154)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.orm.hibernate3.HibernateInterceptor.invoke(HibernateInterceptor.java:111)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:241)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:116)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
	at com.sun.proxy.$Proxy72.doWork(Unknown Source)
	at ome.services.util.Executor$Impl.execute(Executor.java:457)
	at omero.cmd.HandleI.run(HandleI.java:359)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at ome.services.util.Executor$Impl$1.call(Executor.java:498)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:744)
}

	at ome.formats.importer.ImportLibrary$ImportCallback.onFinished(ImportLibrary.java:628)
	at omero.cmd.CmdCallbackI.finished(CmdCallbackI.java:301)
	at omero.cmd._CmdCallbackDisp.___finished(_CmdCallbackDisp.java:118)
	at omero.cmd._CmdCallbackDisp.__dispatch(_CmdCallbackDisp.java:145)
	at IceInternal.Incoming.invoke(Incoming.java:222)
	at Ice.ConnectionI.invokeAll(ConnectionI.java:2482)
	at Ice.ConnectionI.dispatch(ConnectionI.java:1258)
	at Ice.ConnectionI.message(ConnectionI.java:1213)
	at IceInternal.ThreadPool.run(ThreadPool.java:321)
	at IceInternal.ThreadPool.access$300(ThreadPool.java:12)
	at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:693)
	at java.lang.Thread.run(Thread.java:695)

	at org.openmicroscopy.shoola.env.data.util.StatusLabel.update(StatusLabel.java:677)
	at ome.formats.importer.ImportLibrary.notifyObservers(ImportLibrary.java:234)
	at ome.formats.importer.ImportLibrary$ImportCallback.onFinished(ImportLibrary.java:634)
	at omero.cmd.CmdCallbackI.finished(CmdCallbackI.java:301)
	at omero.cmd._CmdCallbackDisp.___finished(_CmdCallbackDisp.java:118)
	at omero.cmd._CmdCallbackDisp.__dispatch(_CmdCallbackDisp.java:145)
	at IceInternal.Incoming.invoke(Incoming.java:222)
	at Ice.ConnectionI.invokeAll(ConnectionI.java:2482)
	at Ice.ConnectionI.dispatch(ConnectionI.java:1258)
	at Ice.ConnectionI.message(ConnectionI.java:1213)
	at IceInternal.ThreadPool.run(ThreadPool.java:321)
	at IceInternal.ThreadPool.access$300(ThreadPool.java:12)
	at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:693)
	at java.lang.Thread.run(Thread.java:695)
Caused by: java.lang.RuntimeException: Failure response on import!
Category: ::omero::grid::ImportRequest
Name: error-on-init
Parameters: {message=For input string: "-merge-deploy/ManagedRepository/user-4_5/2015-01/12/11-36-01", stacktrace=java.lang.NumberFormatException: For input string: "-merge-deploy/ManagedRepository/user-4_5/2015-01/12/11-36-01"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.lang.Integer.parseInt(Integer.java:492)
	at java.lang.Integer.parseInt(Integer.java:527)
	at loci.formats.in.BDReader.initFile(BDReader.java:422)
	at loci.formats.FormatReader.setId(FormatReader.java:1404)
	at loci.formats.ImageReader.setId(ImageReader.java:835)
	at loci.formats.ReaderWrapper.setId(ReaderWrapper.java:651)
	at loci.formats.ChannelFiller.setId(ChannelFiller.java:223)
	at loci.formats.ReaderWrapper.setId(ReaderWrapper.java:651)
	at loci.formats.ChannelSeparator.setId(ChannelSeparator.java:289)
	at loci.formats.ReaderWrapper.setId(ReaderWrapper.java:651)
	at loci.formats.Memoizer.setId(Memoizer.java:627)
	at loci.formats.ReaderWrapper.setId(ReaderWrapper.java:651)
	at ome.services.blitz.repo.CheckedPath.bfSetId(CheckedPath.java:403)
	at ome.services.blitz.repo.ManagedImportRequestI.open(ManagedImportRequestI.java:620)
	at ome.services.blitz.repo.ManagedImportRequestI.init(ManagedImportRequestI.java:225)
	at omero.cmd.HandleI.steps(HandleI.java:427)
	at omero.cmd.HandleI$1.doWork(HandleI.java:365)
	at omero.cmd.HandleI$1.doWork(HandleI.java:361)
	at sun.reflect.GeneratedMethodAccessor293.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at ome.services.util.Executor$Impl$Interceptor.invoke(Executor.java:576)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.security.basic.EventHandler.invoke(EventHandler.java:154)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.orm.hibernate3.HibernateInterceptor.invoke(HibernateInterceptor.java:111)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:241)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:116)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
	at com.sun.proxy.$Proxy72.doWork(Unknown Source)
	at ome.services.util.Executor$Impl.execute(Executor.java:457)
	at omero.cmd.HandleI.run(HandleI.java:359)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at ome.services.util.Executor$Impl$1.call(Executor.java:498)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:744)
}

	at ome.formats.importer.ImportLibrary$ImportCallback.onFinished(ImportLibrary.java:628)
	... 11 more
```

pwalczysko commented:
Graphics in the IJ-importer window:

1. Is it possible to list the images in question (the images to-be-imported) in the list on left-hand side pane like the normal importer does it ? (if not, would it be possible to make the ``Add to the Queue`` button either more prominent or with better description, like ``Add to the Queue the images you selected in ImageJ`` or similar) - I had a short moment of hesitation about what to do actually when the importer window popped up - which button to press, where is my image ?

2. There is some strange "dot" between the single arrow button and double arrow button in the middle of the importer window

![screen shot 2015-01-12 at 11 57 34](https://cloud.githubusercontent.com/assets/2478303/5702510/1a99ada2-9a53-11e4-87c8-09b3a86daf02.png)

pwalczysko commented:
As discussed with @jburel , it should be on the to-do list to have some contingency in case I do not create or select an existing dataset on import - atm this goes to Orphaned - I understand this is planned for a later stage.

jburel commented:
- The import error above is not related to the changes
- The dot has always been there!
- I am going to modify the strategy for windows popping up.

pwalczysko commented:
test2:

  - import of the cropped image failed with the error below (error comes from ImageJ)
  - although there was a ``Failed`` button in the importer, when I click the ``View exception`` in the submenu, nothing happens

```
loci.formats.UnknownFormatException: Unknown file format: /var/folders/5h/4b7kss151v74c0_jgtbt69jh0000gn/T/Screen
	at loci.formats.ImageWriter.getWriter(ImageWriter.java:161)
	at loci.plugins.out.Exporter.run(Exporter.java:281)
	at loci.plugins.LociExporter.run(LociExporter.java:75)
	at ij.plugin.filter.PlugInFilterRunner.processOneImage(PlugInFilterRunner.java:262)
	at ij.plugin.filter.PlugInFilterRunner.<init>(PlugInFilterRunner.java:111)
	at ij.IJ.runUserPlugIn(IJ.java:197)
	at ij.IJ.runPlugIn(IJ.java:160)
	at ij.IJ.runPlugIn(IJ.java:149)
	at org.openmicroscopy.shoola.env.data.model.FileObject.getFileToImport(FileObject.java:148)
	at org.openmicroscopy.shoola.env.data.OmeroImageServiceImpl.importFile(OmeroImageServiceImpl.java:1036)
	at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter.importFile(ImagesImporter.java:78)
	at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter.access$000(ImagesImporter.java:53)
	at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter$1.doCall(ImagesImporter.java:102)
	at org.openmicroscopy.shoola.env.data.views.BatchCall.doStep(BatchCall.java:144)
	at org.openmicroscopy.shoola.util.concur.tasks.CompositeTask.doStep(CompositeTask.java:226)
	at org.openmicroscopy.shoola.env.data.views.CompositeBatchCall.doStep(CompositeBatchCall.java:126)
	at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.exec(ExecCommand.java:165)
	at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.run(ExecCommand.java:276)
	at org.openmicroscopy.shoola.util.concur.tasks.AsyncProcessor$Runner.run(AsyncProcessor.java:91)
	at java.lang.Thread.run(Thread.java:695)
```

pwalczysko commented:
Additional info to the bug https://github.com/openmicroscopy/openmicroscopy/pull/3329#issuecomment-69572836 - this is caused by the presence of spaces of the path of the original image. When ``no_space`` image was tested, the crop imported as expected.

pwalczysko commented:
Bug: Failing to import multi-z crop.

Workflow:
  - open ``test_images_good/leica-liff/Beta Catenin.liff`` in IJ
  - in the Bio-formats importer, select just the last image in the list (the one with multi-z)
  - the image opens as expected, create a rectangular region on it
  - go ``Image > Crop`` in IJ
  - a new stack with cropped images is created
  - import it to OMERO -> fails with

```
loci.formats.UnknownFormatException: Unknown file format: /var/folders/5h/4b7kss151v74c0_jgtbt69jh0000gn/T/Beta
	at loci.formats.ImageWriter.getWriter(ImageWriter.java:161)
	at loci.plugins.out.Exporter.run(Exporter.java:281)
	at loci.plugins.LociExporter.run(LociExporter.java:75)
	at ij.plugin.filter.PlugInFilterRunner.processOneImage(PlugInFilterRunner.java:262)
	at ij.plugin.filter.PlugInFilterRunner.<init>(PlugInFilterRunner.java:111)
	at ij.IJ.runUserPlugIn(IJ.java:197)
	at ij.IJ.runPlugIn(IJ.java:160)
	at ij.IJ.runPlugIn(IJ.java:149)
	at org.openmicroscopy.shoola.env.data.model.FileObject.getFileToImport(FileObject.java:148)
	at org.openmicroscopy.shoola.env.data.OmeroImageServiceImpl.importFile(OmeroImageServiceImpl.java:1036)
	at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter.importFile(ImagesImporter.java:78)
	at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter.access$000(ImagesImporter.java:53)
	at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter$1.doCall(ImagesImporter.java:102)
	at org.openmicroscopy.shoola.env.data.views.BatchCall.doStep(BatchCall.java:144)
	at org.openmicroscopy.shoola.util.concur.tasks.CompositeTask.doStep(CompositeTask.java:226)
	at org.openmicroscopy.shoola.env.data.views.CompositeBatchCall.doStep(CompositeBatchCall.java:126)
	at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.exec(ExecCommand.java:165)
	at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.run(ExecCommand.java:276)
	at org.openmicroscopy.shoola.util.concur.tasks.AsyncProcessor$Runner.run(AsyncProcessor.java:91)
	at java.lang.Thread.run(Thread.java:695)
```

pwalczysko commented:
Created a 3D projection in IJ. Tried to import it to OMERO. This is a 36 planes z stack. After ``Add to the Queue`` nothing happens. (nothing is added to the queue). No error message appears.

pwalczysko commented:
Workflows for single plane images tested (test2):

  - crop -> works as expected
  - translate -> works as expected
  - rotate/flip -> works as expected
  - duplicate -> works as expected
  - background subtraction -> works as expected
  
Workflows for multi-z images tested (test2):

  - crop (multi-z result to import) - fails with error https://github.com/openmicroscopy/openmicroscopy/pull/3329#issuecomment-69598103
  - 3D projection (multi-z result to import) - fails without any error https://github.com/openmicroscopy/openmicroscopy/pull/3329#issuecomment-69607081

pwalczysko commented:
Please note that the multi-z image when opened and imported to OMERO without any changes works just fine. (``Beat catenin.liff``)

bramalingam commented:
crop (multi-z result to import) - fails with error #3329 (comment) : I think, any image size modifications would need a corresponding update in the metadata, which I don't think we are doing currently ... @jburel : please correct me if am wrong here.

bramalingam commented:
Please ignore my previous comment,. I checked the following lines and the width and height are extracted from the ImageJ IP object. 

```
store.setPixelsSizeX(new PositiveInteger(imp.getWidth()), 0);
            store.setPixelsSizeY(new PositiveInteger(imp.getHeight()), 0);
            store.setPixelsSizeZ(new PositiveInteger(imp.getNSlices()), 0);
            store.setPixelsSizeC(new PositiveInteger(channels*imp.getNChannels()), 0);
            store.setPixelsSizeT(new PositiveInteger(imp.getNFrames()), 0);
```

pwalczysko commented:
@jburel : sorry, this https://github.com/openmicroscopy/openmicroscopy/pull/3329#issuecomment-69598103 looks like the spece-in-the-name-problem - will retest the multi-z workflows tomorrow

pwalczysko commented:
@jburel : Even with new ``bioformats_package.jar`` from today's build as well as a new  Insight-ij plugin, I still having the problem which you identified as the space-in-the-name problem, as described in https://github.com/openmicroscopy/openmicroscopy/pull/3329#issuecomment-69572836. This is the same for a single-plane png as well as multi-z liff.

pwalczysko commented:
I can see improvement only on the sequence of opening of importer files. This is as expected (Data Manager opens, then Importer opens over it).

pwalczysko commented:
Retest today:
Working:
  - single plane images
    -  crops saving to OMERO
    - ROIs on crops  drawn in IJ saved to OMERO
  - stacks (single channel liff)
    - crops saving to OMERO
    - flips saving to OMERO

Not working:
  - stacks (single channel liff)
    - ROIs on crops saved to OMERO -> saving works, but ROIs are not displayed on multiple planes in ROI manager, only in the image itself -> leads to mismatch and Insight freezes when trying to select the ROI on another plane

pwalczysko commented:
Further test:
Working:
  - stack with multiple channels when cropped will be imported correctly with correct channels linkage into OMERO

jburel commented:
@pwalczysko: I have added initial support for roi when importing an unmodified image. Only type supported is rectangle. I will add the rest tomorrow.

pwalczysko commented:
Checked 
  - the repeated imports bug with the same name -> works as expected
  - rectangle region on the original image -> works as expected (the only thing is the problem with multi-z, although the region is  propagated in OMERO over all z after import, there is just one shape per region)
  - message is now in importer as expected, will defer to @gusferguson whether this is the best wording and position
  - names with spaces work, although there is a tendency to lose the spaces in the ome.tiff images, e.g. ``Beta catenin.liff`` is ``Beta catenin.liff`` after import to OMERO, but ``Betacatenin.ome.tiff`` after import of a crop to OMERO

pwalczysko commented:
@jburel : 
   - If I go in ImageJ ``OMERO > Connect to OMERO`` then I can in the Data Manager double click onto the thumbnail and it opens in IJ. This is as usual.
   - If I go ``OMERO > Upload and Save to OMERO`` then when I double click on the same image, it opens in OMERO (not in IJ) -> not expected

pwalczysko commented:
This is probably for scoping/mid-term strategy : The two menu items in right-clcik menu on thumb have not the best wording (View in ImageJ means actually to open using Bio-formats). The Open with menu items (IJ and IJ2) do not work for IJ-reasons, but still, they do not work :

Should I create a trello ?

![screen shot 2015-01-14 at 13 16 54](https://cloud.githubusercontent.com/assets/2478303/5739348/99d10266-9bf0-11e4-9801-e8398e20a26e.png)
![screen shot 2015-01-14 at 13 17 04](https://cloud.githubusercontent.com/assets/2478303/5739351/9c92a7d4-9bf0-11e4-9d31-7b4cdabaaa5b.png)

jburel commented:
@pwalczysko: Please add your comments to https://trello.com/c/deNMiFvH/88-plugins-for-imagej-implementation
I will review menu in another PR, wording suggestion welcome cc @gusferguson

pwalczysko commented:
Re https://github.com/openmicroscopy/openmicroscopy/pull/3329#issuecomment-69908874 point 2 (rectangle on original image problem with just one shape in multi-z situation) -> this shows up the same in web, although the region is on every z plane, it is shown as a single shape.

pwalczysko commented:
@jburel : Sorry, I was actually testing multi-t and not a multi-z image for the rectangle on original image in previous comment (corrected them). In case of multi-z image (``test_images_good/dv/...``) I get an error

```
ImageJ 1.47v; Java 1.6.0_65 [64-bit]; Mac OS X 10.8.5; 67MB of 6000MB (1%)
 
java.lang.NullPointerException
	at org.openmicroscopy.shoola.agents.fsimporter.chooser.LocationDialog.sortByUser(LocationDialog.java:1508)
	at org.openmicroscopy.shoola.agents.fsimporter.chooser.LocationDialog.populateLocationComboBoxes(LocationDialog.java:1465)
	at org.openmicroscopy.shoola.agents.fsimporter.chooser.LocationDialog.populateUIWithDisplayData(LocationDialog.java:385)
	at org.openmicroscopy.shoola.agents.fsimporter.chooser.LocationDialog.reset(LocationDialog.java:1596)
	at org.openmicroscopy.shoola.agents.fsimporter.chooser.ImportDialog.reset(ImportDialog.java:1429)
	at org.openmicroscopy.shoola.agents.fsimporter.chooser.ImportDialog.reset(ImportDialog.java:1393)
	at org.openmicroscopy.shoola.agents.fsimporter.view.ImporterComponent.setContainers(ImporterComponent.java:617)
	at org.openmicroscopy.shoola.agents.fsimporter.ImporterAgent.handleExperimenterLoadedDataEvent(ImporterAgent.java:279)
	at org.openmicroscopy.shoola.agents.fsimporter.ImporterAgent.eventFired(ImporterAgent.java:461)
	at org.openmicroscopy.shoola.env.event.EventBusImpl.dispatch(EventBusImpl.java:91)
	at org.openmicroscopy.shoola.env.event.EventBusImpl.post(EventBusImpl.java:222)
	at org.openmicroscopy.shoola.agents.treeviewer.TreeViewerAgent.activate(TreeViewerAgent.java:515)
	at org.openmicroscopy.shoola.env.Container.activateAgents(Container.java:343)
	at org.openmicroscopy.shoola.env.Container.startService(Container.java:376)
	at org.openmicroscopy.shoola.env.init.Initializer.notifyEnd(Initializer.java:197)
	at org.openmicroscopy.shoola.env.Container.startupInPluginMode(Container.java:492)
	at org.openmicroscopy.shoola.env.Container.startupInPluginMode(Container.java:426)
	at org.openmicroscopy.shoola.MainIJPlugin.run(MainIJPlugin.java:262)
	at ij.IJ.runUserPlugIn(IJ.java:195)
	at ij.IJ.runPlugIn(IJ.java:160)
	at ij.Executer.runCommand(Executer.java:128)
	at ij.Executer.run(Executer.java:64)
	at java.lang.Thread.run(Thread.java:695)
```

gusferguson commented:
The View > View icon is the "Open in Full Viewer" icon, so I would expect that menu item to do that. Not sure what you mean about open using Bio-formats.

My choice would be to have a single menu item "View in..." with the following in the submenu:
> Full Viewer
> ImageJ(2)
> ImageJ

jburel commented:
@gusferguson, @pwalczysko: 2 different things that led to the confusion
 * One is "open with " so anything could go under that
 * View: "static" menu entries.

When Opening in ImageJ, we use the B-F plugin to open the image allowing the user to pick various viewing options.

pwalczysko commented:
When I do ``Stack to images`` in IJ and then try to import one of the resutling images into OMERO, I get ``java.io.IOException: No such file or directory``. I guess this will be dealt with as part of the "save everything what is open" effort ?

jburel commented:
@pwalczysko:
 * Added support for all roi types
 * Added support for shapes settings e.g. fill color, stroke width etc.
 * MIF should now be supported
 * Double click on image will open in imageJ, when Importer shows up (second plugin menu option)

jburel commented:
Removing onhold label.

pwalczysko commented:
Still getting
![screen shot 2015-01-19 at 10 38 58](https://cloud.githubusercontent.com/assets/2478303/5799196/725e6522-9fc7-11e4-8552-7f5253a7e22a.png)

pwalczysko commented:
I am not getting any regions when Overlay workflow is used. 
Wofkflow
  - open a .png
  - draw a region, select ``Image > Overlay > Add Selection``
  - verify that the overlay is there by executing ``Hide overlay`` followed by ``Show overlay``
  - select ``Upload and Save to OMERO`` and import the .png to trout
  - no regions found after import

pwalczysko commented:
Still having freezes and/or non-successful imports
  - open .png, create a region using overlay and import to trout (2 x)
  - open ``test_images_good/tiff/Hoechst...`` and draw a region as overlay
  - import the image to trout (the same Project and Dataset as for the 2 previous imports)
  - the importer resports ``File not valid`` (which is not true) and the spinner spins forever, Insight as well as ImageJ freezes
  - note that when you import the ``test_images_good/tiff/Hoechst...`` via ImageJ in a freshly opened ImageJ and fresh connection to trout, it works

pwalczysko commented:
Is it possible that the freeze is some connection lost problem after a long time ? See https://github.com/openmicroscopy/openmicroscopy/pull/3329#issuecomment-70515762 - I could not repeat immediately , but note that the seminar (+1 hour) break took place between the successful and non-successful import in the comment https://github.com/openmicroscopy/openmicroscopy/pull/3329#issuecomment-70515762

jburel commented:
@pwalczysko:
*  Fixed the menu entry
 * For overlay, see screenshot I was able to create overlay by following exactly the same steps See screenshot. I am using 1.47v. Which version of ImageJ are you using.
![overlay](https://cloud.githubusercontent.com/assets/1022396/5808683/2c098c94-a028-11e4-9c37-c8d5de56dbf6.png)

jburel commented:
Followed the steps to import tiff/Hoechst, I have not been able to reproduce. I imported the file few times in a row.

pwalczysko commented:
I have now not been able to repeat https://github.com/openmicroscopy/openmicroscopy/pull/3329#issuecomment-70513243 problem with overlay, in fact, the overlays work fine for many types of images.

pwalczysko commented:
Checked the multi-image import (the radio button in importer on the lower option, see screenshot). Works fine, but RFE: reformulate the text on the lower radio button to read:
``Add images from all image windows`` - I think especially the ``all`` part is important for understnding.

![screen shot 2015-01-20 at 16 39 14](https://cloud.githubusercontent.com/assets/2478303/5821343/38c62bcc-a0c3-11e4-9a4f-c2b0e1b54bef.png)

pwalczysko commented:
Tested:
  - .tiff images -> as expected
  - text change -> as expected
  - overlay in combination with ROI Manager and ROI Manager only -> as expected
  - polyline -> as expected
  - warning message for double menu entry -> gone, as expected
  - the ``Import > to OMERO`` entry -> appearance and function as expected
  - MIF (with different overlay on every MIF member) -> works as expected (although here might be something to think about - when I use ``Add image from all image windows`` I get all the regions right in OMERO, but the MIF will be imported 5x, because it is a five-image-MIF. If I use ``Add image from current window`` I will get the region on the current image on all the imported MIF members and the other regions/overlays on the non-current images are lost)

pwalczysko commented:
I think this is all as expected now. The smaller problems (MIF behaviour, if this can be made better at all, maybe cleaning the message to ``Add images from all image windows``, i.e. adding -s at the end of ``...images...`` etc. ) could be dealt with in another PR (and have somebody else's review as well).

jburel commented:
@pwalczysko: investigating the MIF issue

jburel commented:
@pwalczysko: last commit should fix the MIF import issue

pwalczysko commented:
- open Beta catenin
  - select the Series004 and make a crop
  - do not close anything
  - create an ROI on the cropped Series004 (note that ImageJ changed the name to BetaCatenin.liff - Series004)
  - Save it as Overlay
  - Still in IJ, select image003 and draw a region, save as overlay
  - import to OMERO, selecting "Add from all image windows"
  - full Beta catenin MIF imports 1x (expected) & image003 imports as well separately (not expected)
  - the image003 has the region I have drawn onto the Series004 originally (not expected)

pwalczysko commented:
Additional info to https://github.com/openmicroscopy/openmicroscopy/pull/3329#issuecomment-71034791 - no need to draw any ROIs to reproduce the import problem of the crop. Need just to
   - crop the Series004 (this is the stack)
   - get the Image003 to focus in IJ
   - start the import selecting ``Add from all image windows`` -> the ``*.ome.tiff`` will not be cropped, and will be matching the Image003 -> not expected

pwalczysko commented:
Additional info to https://github.com/openmicroscopy/openmicroscopy/pull/3329#issuecomment-71034791 - the change of focus in the last step is necessary and sufficient condition to reproduce the problem.  The image uploaded as ``.ome-tiff`` is the one which was last in focus. It is also necessary to make the crop from the multi-z stack, if crop is made from the single planes, all works fine.

pwalczysko commented:
Unfortunately, I am able to reproduce the problem also with ``free edge.liff`` from the same folder. 
Again, 
  - select the last image which pops up in the IJ after opening the MIF,
  -  crop it, 
  - get focus into the one-but last (not cropped) image
  - import to omero
  - the ``.ome-tiff`` is identical with the one-but last image in the series (= the one which was the last time in focus)

pwalczysko commented:
Additional info: the problem https://github.com/openmicroscopy/openmicroscopy/pull/3329#issuecomment-71034791 has nothing to do with MIFs, it just came out because we had enough windows open. You can repeat it with a single file/image dv, by creating the crop and reopening the original dv file again (not closing the crop) and then getting the non-cropeed file into focus -> the same problem happens, the ``.ome.tiff`` will not be the cropped image, but the original one (= the one which was in focus).

jburel commented:
@pwalczysko: ok I understood the issue. BF plugin only take into account the current image Plus.
So I need to modify the BF plugin.

jburel commented:
Required https://github.com/openmicroscopy/bioformats/pull/1539

jburel commented:
@pwalczysko: to test the last commit
Click "Quit" before logging in. Then go back to the plugin >OMERO menu. Select one of the entries. The login screen should pop up again (it was not before)

pwalczysko commented:
The problem with the window selection https://github.com/openmicroscopy/openmicroscopy/pull/3329#issuecomment-71052501  is sorted. 
The only follow-up here is the region work on MIFs - unfortunately there is another problem here. The images itself and the crop are all fine, but if I have an ROI on the crop and another ROI on the uncropped original MIF, the ROI on the crop will appear fine in OMERO, the ROI on the MIF will not appear at all (``Beta catenin.liff``).
I think this is a case for another PR, the complex cases like this with MIFs and different regions on different images imported at once can be sorted out later (a simple case of one crop and one non-MIF image, both with different regions, works fine actually, on ``.dv`` file).

pwalczysko commented:
The Quit problem https://github.com/openmicroscopy/openmicroscopy/pull/3329#issuecomment-71198191. 
This does not work. Although there is a second-time splashscreen showing okay, and no crash, there is no action after I login and connect to server.

Workflow:

  - import ``Beta catenin.liff`` using Bio-formats importer (all 5 images)
  - go ``Plugins > OMERO > Upload and Save to OMERO`` -> splashscreen appears as expected
  - click ``Quit``
  - go again ``Plugins > OMERO > Upload and Save to OMERO`` -> splashscreen appears as expected
  - login and connect to server -> nothing happens, neither Data Manager nor importer appear -> not expected

jburel commented:
@pwalczysko:
 * Fix the issue when pressing on "Quit".
 * I have prepared all the code to support overlay on individual MIF. Still need to retrieve the "series" from the OMERO image. Option discussed today with @joshmoore to add a new column "series". So for now the feature will not work.

pwalczysko commented:
The "Quit" problem works as expected.

jburel commented:
@pwalczysko: all the features are now working. The work to handle the crazy MIF scenario is in place, but I will be fully turned on after DB change.
We should merge this PR so we can move forward with the rest and few PRs are conflicting.
Any objections?

sbesson commented:
Got a written agreement from @pwalczysko. Merging

jburel commented:
--no-rebase
